core/src/main/java/com/transferwise/common/gracefulshutdown/strategies/TaskSchedulersGracefulShutdownStrategy.java

```diff
@@ -0,0 +1,73 @@
+package com.transferwise.common.gracefulshutdown.strategies;
+
+import com.transferwise.common.gracefulshutdown.GracefulShutdownStrategy;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicInteger;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationContext;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.scheduling.concurrent.ConcurrentTaskScheduler;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+
+@Slf4j
+public class TaskSchedulersGracefulShutdownStrategy implements GracefulShutdownStrategy {
+
+  @Autowired
+  private ApplicationContext applicationContext;
+
+  private AtomicInteger inProgressShutdowns = new AtomicInteger();
+
+  @Override
+  public void prepareForShutdown() {
+    var executors = Executors.newFixedThreadPool(10);
+
+    var taskSchedulers = applicationContext.getBeansOfType(TaskScheduler.class).values();
+    inProgressShutdowns.getAndSet(taskSchedulers.size());
+
+    for (var taskScheduler : taskSchedulers) {
+      executors.submit(() -> {
+        try {
+          if (taskScheduler instanceof ThreadPoolTaskScheduler) {
+            log.info("Shutting down thread pool task scheduler '{}'.", taskScheduler);
+            var threadPoolTaskScheduler = (ThreadPoolTaskScheduler) taskScheduler;
+            threadPoolTaskScheduler.setWaitForTasksToCompleteOnShutdown(true);
+            threadPoolTaskScheduler.shutdown();
+          } else if (taskScheduler instanceof ConcurrentTaskScheduler) {
+            log.info("Shutting down concurrent task scheduler '{}'.", taskScheduler);
+            var concurrentTaskScheduler = (ConcurrentTaskScheduler) taskScheduler;
+            var executor = concurrentTaskScheduler.getConcurrentExecutor();
+
+            try {
+              var shutdownMethod = executor.getClass().getMethod("shutdown");
+              shutdownMethod.invoke(executor);
+            } catch (NoSuchMethodException ignored) {
+              // ignored
+            } catch (Throwable t) {
+              log.error("Shutting down concurrent task scheduler executor failed.", t);
+            }
+          } else {
+            try {
+              var shutdownMethod = taskScheduler.getClass().getMethod("shutdown");
```

Comment:
This looks odd.  Is `shutdown` guaranteed to exist?  If not, we might need to handle that situation.

Reply:
This is handled by:
```java
} catch (NoSuchMethodException ignored) {
              log.warn("Found a task scheduler '{}', but do not know how to shut it down. Please contact SRE team.", taskScheduler);
            }
```